[pt2] grad support

kshitij12345 opened the pull request:
Teach dynamo about grad

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim

kshitij12345 committed:
[pt2] grad support

kshitij12345 committed:
latest main

kshitij12345 committed:
fixes

kshitij12345 committed:
add test for graph break

kshitij12345 commented on aten/src/ATen/functorch/TensorWrapper.cpp:
```diff
@@ -38,19 +38,27 @@ void dumpTensor(std::ostream& ss, const Tensor& tensor) {
-    sizes_and_strides_.stride_at_unchecked(i) = strides[i];
+  // update size, strides and storage_offset
+  if (value_.unsafeGetTensorImpl()->has_symbolic_sizes_strides()) {
+    // for tensor with symbolic size and strides
```
For support of dynamic shapes.

kshitij12345 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -621,6 +621,154 @@ def fn(x):
+            # Something weird is happening on graph break,
+            # torch.compile(wrapper_fn)(x) returns `torch.func.grad` function
+            # and hence we get an error here.
+            self.assertEqual(actual, expected)
```
Note to self for investigating this.

zou3519 commented on aten/src/ATen/functorch/TensorWrapper.cpp:
```diff
@@ -38,19 +38,27 @@ void dumpTensor(std::ostream& ss, const Tensor& tensor) {
+    // for tensor with symbolic size and strides
+    set_sizes_and_strides(
+        value_.sym_sizes(), value_.sym_strides(), value_.sym_storage_offset());
+  } else {
+    // for tensor with regular size and strides
```
Do we actually need this branch? IIRC `value_.sym_sizes()` is a generalization of `sizes()` that works even if the Tensor isn't symbolic

kshitij12345 replied:
I tried without this branch and all the tests trip at these INTERNAL_ASSERT

https://github.com/pytorch/pytorch/blob/6c9b94dcda696984552769da6cf578ccca0cfbdd/aten/src/ATen/functorch/TensorWrapper.cpp#L162-L172

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1289,6 +1302,99 @@ def speculate_branch(branch):
+                else grad_kwargs.get("argnums", None)
+            )
+
+            # third arg of `grad` is argnums
```
has_aux

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1289,6 +1302,99 @@ def speculate_branch(branch):
+                args,
+                graph_checkpoint,
+                checkpoint,
+            )
```
We need to check that after we performed `speculate_subgraph`, no additional side effects were added. The logic for autograd.Function should have an example of this.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1289,6 +1302,99 @@ def speculate_branch(branch):
+            graph_checkpoint = tx.output.graph
+            grad_args = self.value_args_kwargs["args"]
+            grad_kwargs = self.value_args_kwargs["kwargs"]
+            # first arg of `grad` is function
+            func = grad_args[0] if len(grad_args) >= 1 else grad_kwargs["func"]
```
The default arg handling is not the best. TBH `grad` should probably just invoke either a function or something that subclasses from HigherOrderOperator, and we should have Dynamo inline `grad` so that it sees the thing and captures it. This way we would get the default argument handling for free.

Something like:
```py
def grad(func, argnums=0, has_aux=False):
    return GradHigherOrderOperator(func, argnums, has_aux)
```
or:
```py
def grad(func, argnums=0, has_aux=False):
    return grad_impl(func, argnums, has_aux)
```
where grad_impl is a function that Dynamo tries to lift into a HigherOrderOperator

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1289,6 +1302,99 @@ def speculate_branch(branch):
+            )
+
+            # Pass lifted freevars to the call to `grad_fn`
+            grad_fn_args = tuple(arg.as_proxy() for arg in args) + tuple(
+                body_lifted_freevars
+            )
+            grad_fn_kwargs = {k: v.as_proxy() for k, v in kwargs.items()}
```
Do kwargs actually work? I think speculate_subgraph just assumes your function can't take kwargs?

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1289,6 +1302,99 @@ def speculate_branch(branch):
+            if argnums is not None:
+                if isinstance(argnums.value, int):
+                    example_value = (
+                        args[argnums.value].as_proxy().node.meta["example_value"]
```
The gradients are guaranteed to have the same strides as the input, which could end up mattering. There are two things we could do here:
- literally just execute the forward + backward using FakeTensor to see what the strides of the output are
- add extra .contiguous() nodes to the output, and have the example_value be `args[argnums.value].as_proxy().node.meta["example_value"].contiguous()`

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -621,6 +621,154 @@ def fn(x):
 
+    def _grad_compile_check(self, fn, *inputs):
+        actual = fn(*inputs)
+        expected = torch.compile(fn)(*inputs)
```
testing advice:
- use torch.compile(fn, backend='aot_eager'), we are trying to test that dynamo works (and we don't really care about what inductor does to our code).
- use fullgraph=True. I like this a bit better than checking unique_graphs == 1, but more concretely, this is closer in line with what the rest of this file does
- We should do what the wrap() tests do and check that the graph actually looks like what we expect it to. E.g. there is a GradFunc node somewhere that gets invoked

zou3519 reviewed (commented):
This looks pretty good! I need to run, but I read everything aside from the tests in detail

kshitij12345 committed:
update test

kshitij12345 committed:
updates

kshitij12345 committed:
handle side-effect

kshitij12345 committed:
make linter happy

zou3519 committed:
FIX

kshitij12345 committed:
updates

kshitij12345 committed:
undo unnecessary change

kshitij12345 committed:
fixes

kshitij12345 committed:
Merge branch 'pytorch:main' into grad-torch-compile-support

kshitij12345 committed:
hack import error

kshitij12345 committed:
update

kshitij12345 committed:
latest main

kshitij12345 committed:
handle contiguous and make linter happy

kshitij12345 committed:
update test and comments

kshitij12345 committed:
remove stray comment

kshitij12345 commented on torch/_functorch/eager_transforms.py:
```diff
@@ -1279,112 +1279,15 @@ def wrapper(*args, **kwargs):
-    return wrapper
-
+def grad_impl(func: Callable, argnums: argnums_t, has_aux: bool, args, kwargs):
+    import torch._dynamo
```
Circular import

kshitij12345 committed:
remove stray comment

kshitij12345 committed:
update

kshitij12345 committed:
update

zou3519 commented on torch/_functorch/eager_transforms.py:
```diff
@@ -1279,112 +1283,14 @@ def wrapper(*args, **kwargs):
-        return grad
-    return wrapper
-
+def grad_impl(func: Callable, argnums: argnums_t, has_aux: bool, args, kwargs):
+    func = lazy_dynamo_disable(func)
+    results = grad_and_value(func, argnums, has_aux=has_aux)(*args, **kwargs)
```
We should really make grad_and_value_impl the higher-order operator, so that we get support for both grad and grad_and_value. Doesn't need to happen in this PR.

zou3519 commented on torch/_functorch/utils.py:
```diff
@@ -36,3 +36,12 @@ def wrapper(fn):
     return wrapper
+
+
+@contextlib.contextmanager
+def disable_functorch():
+    guard = torch._C._DisableFuncTorch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
```
you should be able to do:
```
with torch._C._DisableFuncTorch():
     ...
```

zou3519 commented on torch/_functorch/apis.py:
```diff
@@ -0,0 +1,100 @@
+from torch._functorch.eager_transforms import grad_impl, exposed_in, Callable, argnums_t
+
```
Can you add a comment somewhere saying that the reason why we have this file is to workaround limitations in dynamo's skip_files?

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1336,134 @@ def speculate_branch(branch):
+                    - pre_side_effects.id_to_variable.keys()
+                )
+                if len(diff) > 0:
+                    unimplemented("NYI - side effects in torch.func.grad")
```
nit: "NYI - torch.func.grad(f) where there are side effects in f"

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1336,134 @@ def speculate_branch(branch):
+            )
+
+            # Pass lifted freevars to the call to `grad_fn`
+            args = args[3].items
+            grad_fn_args = tuple(arg.as_proxy() for arg in args) + tuple(
+                body_lifted_freevars
+            )
```
Can you test some situations where args:
- are pytrees (both standard and user-defined pytrees)?
- are non-Tensors
I'm not completely convinced this works, so we may need to end up asserting that `args` is a TupleVariable containing TensorVariable.

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1336,134 @@ def speculate_branch(branch):
+            grad_fn_args = tuple(arg.as_proxy() for arg in args) + tuple(
+                body_lifted_freevars
+            )
+            # TODO: kwargs
+            # grad_fn_kwargs = {k: v.as_proxy() for k, v in args[4].items.items()}
```
TODO

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1336,134 @@ def speculate_branch(branch):
+            # grad_output = grad_fn(*grad_fn_args, **grad_fn_kwargs)
+            grad_output = grad_fn(*grad_fn_args)
+
+            # `grad_fn(*grad_fn_args, **grad_fn_kwargs)`
+            # Output of grad_fn is
+            # For has_aux=False, Tuple[gradients of inputs indicated by argnums].
+            # For has_aux=True, Tuple[Tuple[gradients of inputs indicated by argnums], aux values]
```
Yeah, making `grad_and_value_impl` the higher order operator would make it so that we don't need to emulate this logic here in Dynamo; we'd just get Dynamo to trace through it

zou3519 commented on torch/_dynamo/variables/lists.py:
```diff
@@ -230,6 +230,13 @@ def call_method(
+            # For `speculate_subgraph` (HigherOrderOperator), we don't use root tracer,
+            # so here we are banning `append` on list within a `subgraph`.
+            # Ideally, we shouldn't allow dynamo to inline functions under HigherOrderOperator
+            # like `grad`.
```
Link to https://github.com/pytorch/pytorch/issues/103613 please

zou3519 reviewed (commented):
I added a suggestion for the hack, though I want to think about it some more.

kshitij12345 committed:
address review - simplify tensorwrapper refreshmetadata

kshitij12345 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -654,6 +654,229 @@ def fn(x):
+            foo.append(3)
+            return x.sin().sum()
+
+        def wrapper_fn(x):
+            return torch.func.grad(fn)(x)
```
I don't think this will work out of the box now as 
```
torch._dynamo.exc.Unsupported: HigherOrderOperator with body that accepts non-Tensors as input
```

Will add a test to make sure we graph-break for now.

Thanks!

kshitij12345 committed:
update test

kshitij12345 committed:
update test

kshitij12345 committed:
add test

kshitij12345 committed:
address review

kshitij12345 committed:
add config flag

kshitij12345 committed:
add issue reference

kshitij12345 committed:
address review

kshitij12345 committed:
update TODO

kshitij12345 committed:
add comment

kshitij12345 commented:
@zou3519 Have addressed most of the comments (updated tests to use assertExpectedInline, test for pytree and int inputs, updated/added comments, added a config flag)

Two things pending:
* Handling `enable_grad` (waiting for consensus).
* Making `grad_and_value` the higher order operator (would it be ok if I do it in follow-up PR and clean-up `grad` later), lot of things, tests will change to support `grad_and_value` at this point. 

Thank you!

zou3519 commented:
> would it be ok if I do it in follow-up PR and clean-up grad later

Follow-up for that SGTM

kshitij12345 committed:
latest main

kshitij12345 committed:
dont forget functools.wraps

kshitij12345 committed:
fix test

kshitij12345 committed:
move hack to more reasonable place

kshitij12345 committed:
latest main

kshitij12345 committed:
update

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -757,10 +764,27 @@ def safe_or_raise_always_restore(tx, graph_checkpoint, checkpoint, f, sub_args):
 
+@contextlib.contextmanager
+def dynamo_enable_grad(tx):
+    from . import GradModeVariable
+
+    org_value = torch.is_grad_enabled()
+    GradModeVariable.create(tx, True)
+    yield
+    GradModeVariable.create(tx, org_value)
```
Does this need:
```
org_value = torch.is_grad_enabled()
try:
  GradModeVariable.create(tx, True)
  yield
finally:
   GradModeVariable.create(tx, org_value)
```
?
(that's what other context managers using contextlib.contextmanager do, but I'm not sure if we need anything special in this case)

kshitij12345 replied:
Given that we bail out and restore back from checkpoint if there are errors while tracing the function, I thought this would be ok. (But agreed that using with try..finally seems more inline with the codebase, will update this)

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -757,10 +764,27 @@ def safe_or_raise_always_restore(tx, graph_checkpoint, checkpoint, f, sub_args):
+    checkpoint,
+    *,
+    always_restore=False,
+    trace_with_autograd=False,
```
nit: `trace_with_autograd` -> `enable_grad` makes it a bit clearer

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1352,151 @@ def speculate_branch(branch):
             r = body_r.as_proxy().node.meta["example_value"]
             example_value = r
+        elif self.value is torch._functorch.eager_transforms.grad_impl:
+            # TODO: Support `fn` with kwargs.
```
What happens when `fn` has kwargs today? Is it a graph break?

Are you planning on supporting kwargs in this PR or a separate one? Either is fine; if we don't support it in this PR then it should be at least a graph break

kshitij12345 replied:
Adding `graph break` for now and a related test.

zou3519 reviewed (approved):
Code LGTM. I had some minor nits and questions

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1320,6 +1352,151 @@ def speculate_branch(branch):
+            # Output of grad_fn is
+            # For has_aux=False, Tuple[gradients of inputs indicated by argnums].
+            # For has_aux=True, Tuple[Tuple[gradients of inputs indicated by argnums], aux values]
+            # NOTE: example_value should match `grad_output`.
```
I added a section over at https://github.com/pytorch/pytorch/issues/100278

kshitij12345 replied:
Updated thanks!

kshitij12345 committed:
address review

kshitij12345 committed:
latest main

kshitij12345 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -654,6 +665,558 @@ def fn(x):
 
         self.assertTrue(activations.keys() == forward_handles.keys())
 
+    def _grad_compile_check(self, fn, inputs, fullgraph=True):
```
Yes, dynamo shard does have a case for this (that is how we found about this).

kshitij12345 commented:
@zou3519 starting from PR https://github.com/pytorch/pytorch/pull/102310 onwards our hack of graph-breaking on list mutation is leading to failure

Hack Ref:
https://github.com/pytorch/pytorch/blob/ee2854fe47ab204f20db615375ff38f20de39db6/torch/_dynamo/variables/lists.py#L233-L240

One Failing Test (previously it used to trace graph but now it graph-breaks):
https://github.com/pytorch/pytorch/blob/54ff8ffeddab316c43a3ff1f99a1678bdba44a28/test/dynamo/test_higher_order_ops.py#L415

zou3519 commented:
We should delete that hack, then add some expected failures (so that enough tests pass), and worry about it later. It's a pre-existing problem that mutating a list doesn't count as a side effect and we have an issue for it:

kshitij12345 committed:
skip failing dynamo test (mutation not tracked

kshitij12345 commented:
@zou3519 Have used `skipIfTorchDynamo` for now. Couldn't find xfail for dynamo, do you know where it is. Thanks!

kshitij12345 committed:
update expected graph-break msg

zou3519 commented:
> Have used skipIfTorchDynamo for now. Couldn't find xfail for dynamo, do you know where it is? Thanks!

I don't know if that exists haha, skipIfTorchDynamo is fine for now

huydhn commented:
@pytorchbot revert -m 'This is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/85b83954c8820fc7473d8e7b68325fa8ed5753dc and looks like a landrace' -c landrace

kshitij12345 committed:
latest main

kshitij12345 committed:
update test

kshitij12345 commented:
@pytorchbot merge
